File: cpp/113_PathSumII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> ret;
        vector<int> path;
        fetchPath(root,ret,path,sum);
        return ret;
    }
private:
    void fetchPath(TreeNode* root, vector<vector<int>>& ret, vector<int>& path, int sum){
        if(!root) return;
        path.push_back(root->val);
        if(root->val == sum && root->left == NULL && root->right == NULL) ret.push_back(path);
        fetchPath(root->left,ret,path,sum-root->val);
        fetchPath(root->right,ret,path,sum-root->val);
        path.pop_back();
    }
};

File: js/113_PathSumII.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
var pathSum = function(root, sum) {
    var ret = [];
    var path = [];
    fetchPath(root,ret,path,sum);
    return ret;
};

var fetchPath = function(root,ret,path,sum){
    if(!root) return;
    if(root.val === sum && root.left === null && root.right === null) ret.push(path.concat(root.val));
    fetchPath(root.left, ret,path.concat(root.val),sum-root.val);
    fetchPath(root.right,ret,path.concat(root.val),sum-root.val);
};